### PR TITLE
JS SDK construction zone

### PR DESCRIPTION
**Pull request by pro-wh**
for #2

### CODE REVIEW

**Comment by pro-wh** [comment]
current state:

- RPC stuff decodes structures into `Map`s, which is kinda verbose. Making requests with objects as structs works, because they'll encode the same way.
- Void calls incorrectly reject when they succeed. We're waiting for grpc-web to cut a new release with our fix.
- A sample GRPC proxy configuration is included, but it allows all calls, and we've said that exposing the internal GRPC socket is dangerous, so this probably dangerous
- The type definitions are not complete. If we can model every struct and and maintain it, we could add routines to rebuild Maps into objects upon decoding.

**Review by kostko — COMMENTED** [review]
Some general comments:
* We should split the SDK such that there is one package for the general Client SDK (e.g., `@oasisprotocol/client`) and a separate one for the bridge runtime-specific types/queries.
* Similarly, the ledger signer should be in a package that is not bridge-specific.
* We should think about how to best show the difference among the two layers (consensus vs. paratime) in the API, so that it is clear which methods affect which. For example there are accounts and account balances in both the consensus layer and runtimes. Should it be something like "consensus client" vs. "runtime client" (with given runtime ID) constructed over the same node "connection"?
* Would be cool if we could have a transformation step that would encode/decode into proper types based on a schema. Especially for runtimes, at some point we may want to have something like an "introspection query" which would return such a schema so the client would know how to represent specific types. But this is something that can come after the initial API.

**Comment by pro-wh** [comment]
>  * We should split the SDK such that there is one package for the general Client SDK (e.g., `@oasisprotocol/client`) and a separate one for the bridge runtime-specific types/queries.
>
>  * Similarly, the ledger signer should be in a package that is not bridge-specific.

should we get a different repo for this non-bridge general layer? so far everything here is the general part.

>  * We should think about how to best show the difference among the two layers (consensus vs. paratime) in the API, so that it is clear which methods affect which. For example there are accounts and account balances in both the consensus layer and runtimes. Should it be something like "consensus client" vs. "runtime client" (with given runtime ID) constructed over the same node "connection"?

So far I'm not going far down the object-oriented road in this sdk. Currently the separation is clear, which is already a nice thing. They have different name prefixes in the single client and types module, and their helpers are in different modules.

The current design more transparently models the grpc service, which I hope can reduce how much work we need to develop and maintain this package. Bundling a runtime ID with the connection could be slightly convenient, but because it's one parameter, I don't think it offsets the increased cost to learn: "I'm already connected, why can't I make a runtime request? Oh I need a different kind of client for that. How do I get that client? Oh I have to set one up with a runtime ID. Will I have to keep a bunch of clients around if I want to interact with multiple paratimes? Or could I construct and destroy them on the fly? Is the lifecycle for the client heavy? Do I need to explicitly close these clients?" The balance might shift if there are multiple common parameters that we could bundle, or if there's more to the lifecycle that could go wrong.

>  * Would be cool if we could have a transformation step that would encode/decode into proper types based on a schema. Especially for runtimes, at some point we may want to have something like an "introspection query" which would return such a schema so the client would know how to represent specific types. But this is something that can come after the initial API.

I agree that'd be nice. Especially if we could generate it.

**Comment by pro-wh** [comment]
Comments from @lukaw3d:

- We should release a prebuilt version of these libs in js, with .d.ts file. Otherwise it's much harder to set up the compiler for downstream devs.
- Might be an easier way to set up type defs: export example data as JSON, typescript maintains detailed field information, and declare stuff as `<T extends keyof typeof example> ... typeof example[T]`.
- Change some `any`s to `unknown`, where you'd expect devs to know what type they need and cast it.
- Use functional stuff instead of `for` loops. Not gonna switch :stuck_out_tongue: , but he said it, so I'm including it here.

**Comment by kostko** [comment]
> should we get a different repo for this non-bridge general layer? so far everything here is the general part.

Yeah, eventually all of the general stuff (runtime-sdk, client-sdk) will move to a separate repository and this one will only contain the bridge. I was thinking more about the package naming stuff, so the general stuff should probably be called `@oasisprotocol/client` or something like that (and the signers could be `@oasisprotcol/signer-ledger` etc.), not specific to the bridge. Not sure if a bridge-specific package is needed at all (until we have introspection it will probably need to contain some types/helpers).

> So far I'm not going far down the object-oriented road in this sdk. Currently the separation is clear, which is already a nice thing. They have different name prefixes in the single client and types module, and their helpers are in different modules.

Ok, so currently there's no support for runtime stuff, let's see how the API looks once we have that.

> I agree that'd be nice. Especially if we could generate it.

Yeah so my thinking was at least for the runtime part to have the runtime have query methods like `core.GetModuleInfo` which would return metadata for all the modules (information about what queries, callable methods [via transactions], etc. are supported, together with a type schema for the arguments/return types). This could be automatically generated by the runtime SDK from the used types (e.g., via a procedural macro).

The base types would be those that are implemented in the runtime SDK itself and the schema could compose new types from those. Only if a module added very specific types which are not just a composition of existing types they would need to be explicitly supported in the client SDK (or a module companion package).

**Comment by pro-wh** [comment]
I've been adding commits to this. A summary of updates:

- structs are now deserialized into objects
- void calls still incorrectly reject, still waiting for grpc-web upstream
- the sample proxy configuration shows how to whitelist methods--by using a path matcher in the routing config. I don't suppose there's some other sneaky way to call an un-whitelisted method?
- we have so many structs declared now that no `NotModeled` remain

**Comment by pro-wh** [comment]
moved to a non-bridge-specific repo https://github.com/oasisprotocol/oasis-sdk/pull/1 (currently private)

File: client-sdk/ts-web/playground/src/index.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                });
+                blocks.on('data', (block) => {
+                    console.log('block', block);
+                });
+                blocks.on('end', () => {
+                    clearTimeout(cancel);
+                    resolve();
+                });
+            });
+            console.log('done watching');
```
Comment (lukaw3d) [inline client-sdk/ts-web/playground/src/index.js:162]: For fun I'd structure this a bit more artistically

```ts
console.log('watching consensus blocks for 30s');
const blocks = client.consensusWatchBlocks();

blocks.on('status', (status) => {
  console.log('status', status);
});
blocks.on('metadata', (metadata) => {
  console.log('metadata', metadata);
});
blocks.on('data', (block) => {
  console.log('block', block);
});

const watchingPromise = new Promise((resolve, reject) => {
  blocks.on('end', () => resolve());
  blocks.on('error', (e) => reject(e));
});
const timeoutPromise = new Promise((resolve, reject) => {
  setTimeout(() => {
    console.log('time\'s up, cancelling');
    resolve('timeout');
  }, 30_000);
});

try {
  await Promise.race([watchingPromise, timeoutPromise]);
  console.log('done watching');
} finally { // doesn't catch errors
  blocks.cancel(); // hopefully this is ignored if it happens after blocks.on('end')
}
```

or if cancel causes issues after blocks.on('end')
```ts
const timeoutPromise = new Promise((resolve, reject) => {
  setTimeout(() => reject('timeout'), 30_000);
});

try {
  await Promise.race([watchingPromise, timeoutPromise]);
} catch(e) {
  if (e === 'timeout') {
    console.log('time\'s up, cancelling');
    blocks.cancel();
  } else {
    throw e;
  }
}
```

Comment (pro-wh) [reply client-sdk/ts-web/playground/src/index.js:162]: something like this might be nice if we later want to put something in the `finally`

**Pull request closed without merging**